src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs

```diff
@@ -933,54 +914,37 @@ private Task WriteToStreamAsync(ReadOnlyMemory<byte> source, CancellationToken c
 
         private async ValueTask<ArraySegment<byte>> ReadNextLineAsync(CancellationToken cancellationToken)
         {
-            int searchOffset = 0;
+            int previouslyScannedBytes = 0;
             while (true)
             {
-                int startIndex = _readOffset + searchOffset;
-                int length = _readLength - startIndex;
-                int crPos = Array.IndexOf(_readBuffer, (byte)'\r', startIndex, length);
-                if (crPos < 0)
-                {
-                    // Couldn't find a \r.  Read more.
-                    searchOffset = length;
-                    await FillAsync(cancellationToken).ConfigureAwait(false);
-                }
-                else if (crPos + 1 >= _readLength)
+                int scanOffset = _readOffset + previouslyScannedBytes;
+                int endIndex = Array.IndexOf(_readBuffer, (byte)'\n', scanOffset, _readLength - scanOffset);
+                if (endIndex >= 0)
                 {
-                    // We found a \r, but we don't have enough data buffered to read the \n.
-                    searchOffset = length - 1;
-                    await FillAsync(cancellationToken).ConfigureAwait(false);
-                }
-                else if (_readBuffer[crPos + 1] == '\n')
-                {
-                    // We found a \r\n.  Return the data up to and including it.
-                    int lineLength = crPos - _readOffset + 2;
-                    var result = new ArraySegment<byte>(_readBuffer, _readOffset, lineLength);
-                    _readOffset += lineLength;
-                    return result;
-                }
-                else
-                {
-                    ThrowInvalidHttpResponse();
+                    int startIndex = _readOffset;
+                    _readOffset = endIndex + 1;
+
+                    if (endIndex > startIndex && _readBuffer[endIndex - 1] == '\r')
```

Comment:
Just to make sure I understand, the `endIndex > startIndex` check is really just making sure that `endIndex != 0`, right?  I'd personally find the latter easier to understand, as it helps to explain why it's needed before doing the `endIndex - 1` in the subsequent clause.

Reply:
It's checking to make sure endIndex != startIndex.  I can change to that, if you want.

Reply:
To be clear, it's not a check for endIndex != 0 or endIndex > 0.  It's a check for (endIndex - startIndex) != 0 or (endIndex - startIndex) == 0.

I do want the logic here to be as clear as possible.  We had an issue here previously.  The endIndex/startIndex stuff was my attempt to clarify, but apparently not very successful :)

An alternative would be to use something like "length" instead of "endIndex"; then the test becomes length != 0, which may be the clearest way to handle this...

Thoughts?

Reply:
We know that endIndex points to \n, or we wouldn't be here.  Thus all we care about in this condition is whether it's preceded by a \r.  Either this is the first char in the array (endIndex == 0), in which case it's not preceded by a \r, or this is later in the array, and we can check endIndex-1 to see if it's \r.  That's all I'm saying :)